File: algorithms/graph_algorithms/dijkstra/dijkstra.rb
# Dijkstra's Algorithm: Shortest path in weighted graphs
require 'set'


def dijkstra_heap(graph, start, goal)
  pq = [[0, start, [start]]]
  visited = Set.new
  
  until pq.empty?
    cost, node, path = pq.min_by { |c, _, _| c }
    pq.delete([cost, node, path])
    
    next if visited.include?(node)
    return [path, cost] if node == goal
    
    visited.add(node)
    (graph[node] || []).each do |neighbor, weight|
      unless visited.include?(neighbor)
        pq << [cost + weight, neighbor, path + [neighbor]]
      end
    end
  end
  
  [nil, Float::INFINITY]
end


def dijkstra_array(graph, start, goal)
  distances = Hash.new(Float::INFINITY)
  graph.keys.each { |n| distances[n] = Float::INFINITY }
  distances[start] = 0
  previous = {}
  unvisited = graph.keys.to_set
  
  until unvisited.empty?
    current = unvisited.min_by { |n| distances[n] }
    
    break if distances[current] == Float::INFINITY
    
    if current == goal
      path = []
      node = current
      while node
        path.unshift(node)
        node = previous[node]
      end
      return [path, distances[goal]]
    end
    
    unvisited.delete(current)
    
    (graph[current] || []).each do |neighbor, weight|
      new_dist = distances[current] + weight
      if new_dist < distances[neighbor]
        distances[neighbor] = new_dist
        previous[neighbor] = current
      end
    end
  end
  
  [nil, Float::INFINITY]
end


def dijkstra_all_paths(graph, start)
  distances = Hash.new(Float::INFINITY)
  graph.keys.each { |n| distances[n] = Float::INFINITY }
  distances[start] = 0
  pq = [[0, start]]
  visited = Set.new
  
  until pq.empty?
    cost, node = pq.min_by { |c, _| c }
    pq.delete([cost, node])
    
    next if visited.include?(node)
    visited.add(node)
    
    (graph[node] || []).each do |neighbor, weight|
      new_dist = cost + weight
      if new_dist < distances[neighbor]
        distances[neighbor] = new_dist
        pq << [new_dist, neighbor]
      end
    end
  end
  
  distances
end

File: algorithms/graph_algorithms/dijkstra/dijkstra.js
// Dijkstra's Algorithm: Shortest path in weighted graphs


function dijkstra_heap(graph, start, goal) {
    const pq = [[0, start, [start]]];
    const visited = new Set();
    
    while (pq.length > 0) {
        pq.sort((a, b) => a[0] - b[0]);
        const [cost, node, path] = pq.shift();
        
        if (visited.has(node)) continue;
        if (node === goal) return [path, cost];
        
        visited.add(node);
        for (const [neighbor, weight] of (graph[node] || [])) {
            if (!visited.has(neighbor)) {
                pq.push([cost + weight, neighbor, [...path, neighbor]]);
            }
        }
    }
    
    return [null, Infinity];
}


function dijkstra_array(graph, start, goal) {
    const distances = {};
    const previous = {};
    const unvisited = new Set();
    
    for (const node in graph) {
        distances[node] = Infinity;
        unvisited.add(node);
    }
    distances[start] = 0;
    
    while (unvisited.size > 0) {
        let current = null;
        let minDist = Infinity;
        for (const node of unvisited) {
            if (distances[node] < minDist) {
                minDist = distances[node];
                current = node;
            }
        }
        
        if (minDist === Infinity) break;
        
        if (current === goal) {
            const path = [];
            let node = current;
            while (node) {
                path.unshift(node);
                node = previous[node];
            }
            return [path, distances[goal]];
        }
        
        unvisited.delete(current);
        
        for (const [neighbor, weight] of (graph[current] || [])) {
            const newDist = distances[current] + weight;
            if (newDist < distances[neighbor]) {
                distances[neighbor] = newDist;
                previous[neighbor] = current;
            }
        }
    }
    
    return [null, Infinity];
}


function dijkstra_all_paths(graph, start) {
    const distances = {};
    for (const node in graph) {
        distances[node] = Infinity;
    }
    distances[start] = 0;
    
    const pq = [[0, start]];
    const visited = new Set();
    
    while (pq.length > 0) {
        pq.sort((a, b) => a[0] - b[0]);
        const [cost, node] = pq.shift();
        
        if (visited.has(node)) continue;
        visited.add(node);
        
        for (const [neighbor, weight] of (graph[node] || [])) {
            const newDist = cost + weight;
            if (newDist < distances[neighbor]) {
                distances[neighbor] = newDist;
                pq.push([newDist, neighbor]);
            }
        }
    }
    
    return distances;
}


module.exports = { dijkstra_heap, dijkstra_array, dijkstra_all_paths };

